doc/ips.md

```diff
@@ -0,0 +1,323 @@
+
+Each peerlist is calculated over the original network. Due to this, it is proposed to limit the number of peers changed for each node at once, in order to limit the influence of potential bad decisions on the network. Changing a small amount of nodes at once creates a more evolutionary approach to network shaping.
+
+The algorithm code is in `ips/algorithm.rs` and is extensively commented.
+
+### Factors
+The following factors are taken into account when talking about the network:
++ degree – a count, representing how many direct, ‘one hop’ connections each node has to other nodes in the network. IPS is trying to keep this measurement close to the network average for degree, and attempt to construct a peerlist with neighbor counts somewhere close to the middle between the current and average degrees.
++ betweenness – broadly, this tells us us how often a node lies on a path between other network nodes. It is computed by identifying all the shortest paths and then counting how many times each node falls on one. IPS tries to keep it close to the network average for the node; node should search peers with a high betweenness value as it means that the peer is often on the shortest path.
++ eigenvector – this tells how much influence node's neighbours have.
++ closeness – this measure calculates the shortest paths between all nodes, then assigns each node a score based on its sum of shortest paths. This is not very relevant here, as neither the density or sparseness of a network is intrinsically bad. IPS tries to keep its own centrality high and connect to peers with high closeness (if the MCDA weights allow it).
+
+All of the above factors are written to the IPS result log, allowing one to check how a particular run influenced the network: the state before may be compared to the the state afterwards.
+
+### Security checks
+
+One of the most important properties of a network topology is the presence or absence of network islands. Presence of islands may influence every other network or node parameter that we are observing. Merging two massive islands can be risky and should not be done automatically. The islands could have been disconnected for a long time and produced different history of their blockchain. IPS can detect such situation and breaks computations notifying the user about island existence. 
+
+Next, IPS checks if the network can be easily fragmented by attacking a given percent of the nodes and if so, preventing such cases by creating new connections between their neighbours. Nodes selected to that fragmentation simulation are chosen from the "hot" nodes, which means the nodes with highest betweenness factor. 
```

Comment:
`Nodes selected by that fragmentation simulation`

Reply:
👍